archive/100-theorems-list/93_birthday_problem.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import data.fintype.card
+import data.nat.factorial
+import tactic
+
+/--!
+# Birthday Problem
+
+This file proves Theorem 93 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+
+As of writing, probability is not implemented in Lean, so we instead state the birthday problem
+in terms of injective functions. We instead prove a general result about ‖α ↪ β‖.
+-/
+
+open_locale classical nat
+
+open finset function nat
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation. -/
+def desc_fac (n : ℕ) : ℕ → ℕ
```

Comment:
I thought I saw @semorrison add this as a definition somewhere else, but might be misremembering.

Reply:
If you're thinking of [`pochhammer`](https://leanprover-community.github.io/mathlib_docs/ring_theory/polynomial/pochhammer.html#pochhammer), it's stated in terms of polys and needs `eval` and stuff

Reply:
Could still switch over if that's better though - I worry it won't be fast enough for `norm_num`